### ISSUE TITLE
License Generation Issue: Invalid License Error

### ISSUE THREAD

**Issue by 4S7xfPcbp1Hfayz** [rrrrrrri/fos-license-gen#14]
Hello,

I've used your tool, but I didn't succeed. I believe I may have missed something. Here are the details of my FortiGate:

```
FortiFirewall-VM64-KVM # diagnose debug vm-print-license
SerialNumber: FFVMEVNKHXAHC035
CreateDate: Mon Dec 18 06:23:29 2023
Model: EVAL (1)
CPU: 1
MEM: 2048
VDOM license:
  permanent: 2
  subscription: 0

FortiFirewall-VM64-KVM #   get system status
Version: FortiFirewall-VM64-KVM v7.4.1,build2463,230830 (GA.F)
Security Level: 1
Secure Boot: Disabled
Serial-Number: FFVMEVNKHXAHC035
License Status: Invalid
VM Resources: 1 CPU/1 allowed, 1992 MB RAM/2048 MB allowed
Log hard disk: Available
Hostname: FortiFirewall-VM64-KVM
Operation Mode: NAT
Current virtual domain: root
Max number of virtual domains: 2
Virtual domains status: 1 in NAT mode, 0 in TP mode
Virtual domain configuration: disable
FIPS-CC mode: disable
Current HA mode: standalone
Branch point: 2463
Release Version Information: GA
FortiOS x86-64: Yes
System time: Mon Dec 18 06:29:22 2023
Last reboot reason: power cycle
```

Here are the modifications I made to generate the license:

```
LicenseDataBlock("SERIALNO", "FGVMPGFFVMEVNKHXAHC035"),
LicenseDataBlock("CREATEDATE", "1702910078"),
```

I used the tool https://www.epochconverter.com/ to obtain today's date.

Can you please let me know if something is missing because when I import the license, I receive an error message indicating that it's invalid?

**Comment by 4S7xfPcbp1Hfayz** [comment]
Thank you for your response. I also tried 7.2.6, but it doesn't work. Could you please tell me what I need to change in the LicenseDataBlock?

```
FortiFirewall-VM64-KVM # diagnose debug vm-print-license
SerialNumber: FFVMEVNKHXAHC035
CreateDate: Mon Dec 18 06:56:49 2023
Model: EVAL (1)
CPU: 1
MEM: 2048
VDOM license:
  permanent: 2
  subscription: 0

FortiFirewall-VM64-KVM # get system status
Version: FortiFirewall-VM64-KVM v7.2.6,build1575,230926 (GA.F)
Security Level: 1
Serial-Number: FFVMEVNKHXAHC035
License Status: Invalid
VM Resources: 1 CPU/1 allowed, 1993 MB RAM/2048 MB allowed
Log hard disk: Available
Hostname: FortiFirewall-VM64-KVM
Operation Mode: NAT
Current virtual domain: root
Max number of virtual domains: 2
Virtual domains status: 1 in NAT mode, 0 in TP mode
Virtual domain configuration: disable
FIPS-CC mode: disable
Current HA mode: standalone
Branch point: 1575
Release Version Information: GA
FortiOS x86-64: Yes
System time: Mon Dec 18 06:59:07 2023
Last reboot reason: power cycle
```

**Comment by 4S7xfPcbp1Hfayz** [comment]
I just tried with FGT_VM64_KVM-v7.2.4.F-build1396-FORTINET.out.kvm.zip, and it works! 
Before, I had tested with FFW_VM64_KVM-v7.2.6.F-build1575-FORTINET.out.kvm.zip

**Comment by 4S7xfPcbp1Hfayz** [comment]
I just ran some additional tests, and it turns out it works with FortiGate 7.4.X but not with FortiFirewall. I had selected the wrong file.

FortiFirewall (FFW) vs. FortiGate (FGT):
- FortiFirewall prioritizes throughput and does not have any UTM features. It is mainly used by ISPs or telecommunications customers to maximize bandwidth for consumer internet traffic. Therefore, the primary differences between each license tier can be measured in terms of bandwidth.

- FortiGate prioritizes network security and is popular among businesses to prevent unauthorized access to internal resources. The associated licenses are mostly for hardware features to enhance network security.

I tested it this morning, and it works with the following:

FGT_VM64_KVM-v7.4.1.F-build2463-FORTINET.out.kvm.zip
f65702de07ed1ecbd442f7711fe68f46 (Regular)
3e81a61108e1814ac7300aa494302e37738de30567574cb43aae41226d566bd027e61ec9c586fc4bf4b17bba565b1aba676756015f8e5f8d04293b99e5942315 (SHA-512)

FGT_VM64_KVM-v7.2.6.F-build1575-FORTINET.out.kvm.zip
e5ff042d9f5afd32bfdb4d51436577d5 (Regular)
8c202bf7d6de90fa35e139f771449c044ed5cb31802605dadaed57cdbc8c2e281cbfccd181cd7603c2e915232a3249f19e111115a1d61e4af7c9940a03f9dd5f (SHA-512)

**Comment by rrrrrrri** [comment]
> I just ran some additional tests, and it turns out it works with FortiGate 7.4.X but not with FortiFirewall. I had selected the wrong file.
> 
> FortiFirewall (FFW) vs. FortiGate (FGT):
> 
> * FortiFirewall prioritizes throughput and does not have any UTM features. It is mainly used by ISPs or telecommunications customers to maximize bandwidth for consumer internet traffic. Therefore, the primary differences between each license tier can be measured in terms of bandwidth.
> * FortiGate prioritizes network security and is popular among businesses to prevent unauthorized access to internal resources. The associated licenses are mostly for hardware features to enhance network security.
> 
> I tested it this morning, and it works with the following:
> 
> FGT_VM64_KVM-v7.4.1.F-build2463-FORTINET.out.kvm.zip f65702de07ed1ecbd442f7711fe68f46 (Regular) 3e81a61108e1814ac7300aa494302e37738de30567574cb43aae41226d566bd027e61ec9c586fc4bf4b17bba565b1aba676756015f8e5f8d04293b99e5942315 (SHA-512)
> 
> FGT_VM64_KVM-v7.2.6.F-build1575-FORTINET.out.kvm.zip e5ff042d9f5afd32bfdb4d51436577d5 (Regular) 8c202bf7d6de90fa35e139f771449c044ed5cb31802605dadaed57cdbc8c2e281cbfccd181cd7603c2e915232a3249f19e111115a1d61e4af7c9940a03f9dd5f (SHA-512)

This tool only supports FortiGate Esxi/Vmware 7.2/7.4 versions, others are not tested at this time.

**Comment by 4S7xfPcbp1Hfayz** [comment]
Yes, I understood, but FortiGate is the only thing that interests me anyway. I would also like some explanations about  
```
LicenseDataBlock("USGFACTORY", "0"),
LicenseDataBlock("LENCFACTORY", "0"),
LicenseDataBlock("CARRIERFACTORY", "0")
````
because I don't understand their purpose.

**Comment by rrrrrrri** [comment]
> Yes, I understood, but FortiGate is the only thing that interests me anyway. I would also like some explanations about
> 
> ```
> LicenseDataBlock("USGFACTORY", "0"),
> LicenseDataBlock("LENCFACTORY", "0"),
> LicenseDataBlock("CARRIERFACTORY", "0")
> ```
> 
> because I don't understand their purpose.

These flag bits may exist in the License to identify whether the system is in USG/LENC/CARRIER mode.

The License generated by this tool does not contain certain information. It is recommended to use it only for security testing purposes and not for production environments.
